// @redis/search/dist/lib/commands/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const _LIST_1 = __importDefault(require("./_LIST"));
const ALTER_1 = __importDefault(require("./ALTER"));
const AGGREGATE_WITHCURSOR_1 = __importDefault(require("./AGGREGATE_WITHCURSOR"));
const AGGREGATE_1 = __importDefault(require("./AGGREGATE"));
const ALIASADD_1 = __importDefault(require("./ALIASADD"));
const ALIASDEL_1 = __importDefault(require("./ALIASDEL"));
const ALIASUPDATE_1 = __importDefault(require("./ALIASUPDATE"));
const CONFIG_GET_1 = __importDefault(require("./CONFIG_GET"));
const CONFIG_SET_1 = __importDefault(require("./CONFIG_SET"));
const CREATE_1 = __importDefault(require("./CREATE"));
const CURSOR_DEL_1 = __importDefault(require("./CURSOR_DEL"));
const CURSOR_READ_1 = __importDefault(require("./CURSOR_READ"));
const DICTADD_1 = __importDefault(require("./DICTADD"));
const DICTDEL_1 = __importDefault(require("./DICTDEL"));
const DICTDUMP_1 = __importDefault(require("./DICTDUMP"));
const DROPINDEX_1 = __importDefault(require("./DROPINDEX"));
const EXPLAIN_1 = __importDefault(require("./EXPLAIN"));
const EXPLAINCLI_1 = __importDefault(require("./EXPLAINCLI"));
const HYBRID_1 = __importDefault(require("./HYBRID"));
const INFO_1 = __importDefault(require("./INFO"));
const PROFILE_SEARCH_1 = __importDefault(require("./PROFILE_SEARCH"));
const PROFILE_AGGREGATE_1 = __importDefault(require("./PROFILE_AGGREGATE"));
const SEARCH_NOCONTENT_1 = __importDefault(require("./SEARCH_NOCONTENT"));
const SEARCH_1 = __importDefault(require("./SEARCH"));
const SPELLCHECK_1 = __importDefault(require("./SPELLCHECK"));
const SUGADD_1 = __importDefault(require("./SUGADD"));
const SUGDEL_1 = __importDefault(require("./SUGDEL"));
const SUGGET_WITHPAYLOADS_1 = __importDefault(require("./SUGGET_WITHPAYLOADS"));
const SUGGET_WITHSCORES_WITHPAYLOADS_1 = __importDefault(require("./SUGGET_WITHSCORES_WITHPAYLOADS"));
const SUGGET_WITHSCORES_1 = __importDefault(require("./SUGGET_WITHSCORES"));
const SUGGET_1 = __importDefault(require("./SUGGET"));
const SUGLEN_1 = __importDefault(require("./SUGLEN"));
const SYNDUMP_1 = __importDefault(require("./SYNDUMP"));
const SYNUPDATE_1 = __importDefault(require("./SYNUPDATE"));
const TAGVALS_1 = __importDefault(require("./TAGVALS"));
exports.default = {
    /**
     * Lists all existing indexes in the database.
     */
    _LIST: _LIST_1.default,
    /**
     * Lists all existing indexes in the database.
     */
    _list: _LIST_1.default,
    /**
     * Alters an existing RediSearch index schema by adding new fields.
     * @param index - The index to alter
     * @param schema - The schema definition containing new fields to add
     */
    ALTER: ALTER_1.default,
    /**
     * Alters an existing RediSearch index schema by adding new fields.
     * @param index - The index to alter
     * @param schema - The schema definition containing new fields to add
     */
    alter: ALTER_1.default,
    /**
     * Performs an aggregation with a cursor for retrieving large result sets.
     * @param index - Name of the index to query
     * @param query - The aggregation query
     * @param options - Optional parameters:
     *   - All options supported by FT.AGGREGATE
     *   - COUNT: Number of results to return per cursor fetch
     *   - MAXIDLE: Maximum idle time for cursor in milliseconds
     */
    AGGREGATE_WITHCURSOR: AGGREGATE_WITHCURSOR_1.default,
    /**
     * Performs an aggregation with a cursor for retrieving large result sets.
     * @param index - Name of the index to query
     * @param query - The aggregation query
     * @param options - Optional parameters:
     *   - All options supported by FT.AGGREGATE
     *   - COUNT: Number of results to return per cursor fetch
     *   - MAXIDLE: Maximum idle time for cursor in milliseconds
     */
    aggregateWithCursor: AGGREGATE_WITHCURSOR_1.default,
    /**
     * Performs an aggregation query on a RediSearch index.
     * @param index - The index name to query
     * @param query - The text query to use as filter, use * to indicate no filtering
     * @param options - Optional parameters for aggregation:
     *   - VERBATIM: disable stemming in query evaluation
     *   - LOAD: specify fields to load from documents
     *   - STEPS: sequence of aggregation steps (GROUPBY, SORTBY, APPLY, LIMIT, FILTER)
     *   - PARAMS: bind parameters for query evaluation
     *   - TIMEOUT: maximum time to run the query
     */
    AGGREGATE: AGGREGATE_1.default,
    /**
     * Performs an aggregation query on a RediSearch index.
     * @param index - The index name to query
     * @param query - The text query to use as filter, use * to indicate no filtering
     * @param options - Optional parameters for aggregation:
     *   - VERBATIM: disable stemming in query evaluation
     *   - LOAD: specify fields to load from documents
     *   - STEPS: sequence of aggregation steps (GROUPBY, SORTBY, APPLY, LIMIT, FILTER)
     *   - PARAMS: bind parameters for query evaluation
     *   - TIMEOUT: maximum time to run the query
     */
    aggregate: AGGREGATE_1.default,
    /**
     * Adds an alias to a RediSearch index.
     * @param alias - The alias to add
     * @param index - The index name to alias
     */
    ALIASADD: ALIASADD_1.default,
    /**
     * Adds an alias to a RediSearch index.
     * @param alias - The alias to add
     * @param index - The index name to alias
     */
    aliasAdd: ALIASADD_1.default,
    /**
     * Removes an existing alias from a RediSearch index.
     * @param alias - The alias to remove
     */
    ALIASDEL: ALIASDEL_1.default,
    /**
     * Removes an existing alias from a RediSearch index.
     * @param alias - The alias to remove
     */
    aliasDel: ALIASDEL_1.default,
    /**
     * Updates the index pointed to by an existing alias.
     * @param alias - The existing alias to update
     * @param index - The new index name that the alias should point to
     */
    ALIASUPDATE: ALIASUPDATE_1.default,
    /**
     * Updates the index pointed to by an existing alias.
     * @param alias - The existing alias to update
     * @param index - The new index name that the alias should point to
     */
    aliasUpdate: ALIASUPDATE_1.default,
    /**
     * Gets a RediSearch configuration option value.
     * @param option - The name of the configuration option to retrieve
     */
    CONFIG_GET: CONFIG_GET_1.default,
    /**
     * Gets a RediSearch configuration option value.
     * @param option - The name of the configuration option to retrieve
     */
    configGet: CONFIG_GET_1.default,
    /**
     * Sets a RediSearch configuration option value.
     * @param property - The name of the configuration option to set
     * @param value - The value to set for the configuration option
     */
    CONFIG_SET: CONFIG_SET_1.default,
    /**
     * Sets a RediSearch configuration option value.
     * @param property - The name of the configuration option to set
     * @param value - The value to set for the configuration option
     */
    configSet: CONFIG_SET_1.default,
    /**
     * Creates a new search index with the given schema and options.
     * @param index - Name of the index to create
     * @param schema - Index schema defining field names and types (TEXT, NUMERIC, GEO, TAG, VECTOR, GEOSHAPE).
     *   Each field can be a single definition or an array to index the same field multiple times with different configurations.
     * @param options - Optional parameters:
     *   - ON: Type of container to index (HASH or JSON)
     *   - PREFIX: Prefixes for document keys to index
     *   - FILTER: Expression that filters indexed documents
     *   - LANGUAGE/LANGUAGE_FIELD: Default language for indexing
     *   - SCORE/SCORE_FIELD: Document ranking parameters
     *   - MAXTEXTFIELDS: Index all text fields without specifying them
     *   - TEMPORARY: Create a temporary index
     *   - NOOFFSETS/NOHL/NOFIELDS/NOFREQS: Index optimization flags
     *   - STOPWORDS: Custom stopword list
     */
    CREATE: CREATE_1.default,
    /**
     * Creates a new search index with the given schema and options.
     * @param index - Name of the index to create
     * @param schema - Index schema defining field names and types (TEXT, NUMERIC, GEO, TAG, VECTOR, GEOSHAPE).
     *   Each field can be a single definition or an array to index the same field multiple times with different configurations.
     * @param options - Optional parameters:
     *   - ON: Type of container to index (HASH or JSON)
     *   - PREFIX: Prefixes for document keys to index
     *   - FILTER: Expression that filters indexed documents
     *   - LANGUAGE/LANGUAGE_FIELD: Default language for indexing
     *   - SCORE/SCORE_FIELD: Document ranking parameters
     *   - MAXTEXTFIELDS: Index all text fields without specifying them
     *   - TEMPORARY: Create a temporary index
     *   - NOOFFSETS/NOHL/NOFIELDS/NOFREQS: Index optimization flags
     *   - STOPWORDS: Custom stopword list
     */
    create: CREATE_1.default,
    /**
     * Deletes a cursor from an index.
     * @param index - The index name that contains the cursor
     * @param cursorId - The cursor ID to delete
     */
    CURSOR_DEL: CURSOR_DEL_1.default,
    /**
     * Deletes a cursor from an index.
     * @param index - The index name that contains the cursor
     * @param cursorId - The cursor ID to delete
     */
    cursorDel: CURSOR_DEL_1.default,
    /**
     * Reads from an existing cursor to get more results from an index.
     * @param index - The index name that contains the cursor
     * @param cursor - The cursor ID to read from
     * @param options - Optional parameters:
     *   - COUNT: Maximum number of results to return
     */
    CURSOR_READ: CURSOR_READ_1.default,
    /**
     * Reads from an existing cursor to get more results from an index.
     * @param index - The index name that contains the cursor
     * @param cursor - The cursor ID to read from
     * @param options - Optional parameters:
     *   - COUNT: Maximum number of results to return
     */
    cursorRead: CURSOR_READ_1.default,
    /**
     * Adds terms to a dictionary.
     * @param dictionary - Name of the dictionary to add terms to
     * @param term - One or more terms to add to the dictionary
     */
    DICTADD: DICTADD_1.default,
    /**
     * Adds terms to a dictionary.
     * @param dictionary - Name of the dictionary to add terms to
     * @param term - One or more terms to add to the dictionary
     */
    dictAdd: DICTADD_1.default,
    /**
     * Deletes terms from a dictionary.
     * @param dictionary - Name of the dictionary to remove terms from
     * @param term - One or more terms to delete from the dictionary
     */
    DICTDEL: DICTDEL_1.default,
    /**
     * Deletes terms from a dictionary.
     * @param dictionary - Name of the dictionary to remove terms from
     * @param term - One or more terms to delete from the dictionary
     */
    dictDel: DICTDEL_1.default,
    /**
     * Returns all terms in a dictionary.
     * @param dictionary - Name of the dictionary to dump
     */
    DICTDUMP: DICTDUMP_1.default,
    /**
     * Returns all terms in a dictionary.
     * @param dictionary - Name of the dictionary to dump
     */
    dictDump: DICTDUMP_1.default,
    /**
     * Deletes an index and all associated documents.
     * @param index - Name of the index to delete
     * @param options - Optional parameters:
     *   - DD: Also delete the indexed documents themselves
     */
    DROPINDEX: DROPINDEX_1.default,
    /**
     * Deletes an index and all associated documents.
     * @param index - Name of the index to delete
     * @param options - Optional parameters:
     *   - DD: Also delete the indexed documents themselves
     */
    dropIndex: DROPINDEX_1.default,
    /**
     * Returns the execution plan for a complex query.
     * @param index - Name of the index to explain query against
     * @param query - The query string to explain
     * @param options - Optional parameters:
     *   - PARAMS: Named parameters to use in the query
     *   - DIALECT: Version of query dialect to use (defaults to 1)
     */
    EXPLAIN: EXPLAIN_1.default,
    /**
     * Returns the execution plan for a complex query.
     * @param index - Name of the index to explain query against
     * @param query - The query string to explain
     * @param options - Optional parameters:
     *   - PARAMS: Named parameters to use in the query
     *   - DIALECT: Version of query dialect to use (defaults to 1)
     */
    explain: EXPLAIN_1.default,
    /**
     * Returns the execution plan for a complex query in a more verbose format than FT.EXPLAIN.
     * @param index - Name of the index to explain query against
     * @param query - The query string to explain
     * @param options - Optional parameters:
     *   - DIALECT: Version of query dialect to use (defaults to 1)
     */
    EXPLAINCLI: EXPLAINCLI_1.default,
    /**
     * Returns the execution plan for a complex query in a more verbose format than FT.EXPLAIN.
     * @param index - Name of the index to explain query against
     * @param query - The query string to explain
     * @param options - Optional parameters:
     *   - DIALECT: Version of query dialect to use (defaults to 1)
     */
    explainCli: EXPLAINCLI_1.default,
    /**
     * Performs a hybrid search combining multiple search expressions.
     * Supports multiple SEARCH and VECTOR expressions with various fusion methods.
     *
     * @experimental
     * NOTE: FT.Hybrid is still in experimental state
     * It's behaviour and function signature may change
     *
     * @param index - The index name to search
     * @param options - Hybrid search options including:
     *   - SEARCH: Text search expression with optional scoring
     *   - VSIM: Vector similarity expression with KNN/RANGE methods
     *   - COMBINE: Fusion method (RRF, LINEAR)
     *   - Post-processing operations: LOAD, GROUPBY, APPLY, SORTBY, FILTER
     *   - Tunable options: LIMIT, PARAMS, TIMEOUT
     */
    HYBRID: HYBRID_1.default,
    /**
     * Performs a hybrid search combining multiple search expressions.
     * Supports multiple SEARCH and VECTOR expressions with various fusion methods.
     *
     * @experimental
     * NOTE: FT.Hybrid is still in experimental state
     * It's behaviour and function signature may change
     *
     * @param index - The index name to search
     * @param options - Hybrid search options including:
     *   - SEARCH: Text search expression with optional scoring
     *   - VSIM: Vector similarity expression with KNN/RANGE methods
     *   - COMBINE: Fusion method (RRF, LINEAR)
     *   - Post-processing operations: LOAD, GROUPBY, APPLY, SORTBY, FILTER
     *   - Tunable options: LIMIT, PARAMS, TIMEOUT
     */
    hybrid: HYBRID_1.default,
    /**
     * Returns information and statistics about an index.
     * @param index - Name of the index to get information about
     */
    INFO: INFO_1.default,
    /**
     * Returns information and statistics about an index.
     * @param index - Name of the index to get information about
     */
    info: INFO_1.default,
    /**
     * Profiles the execution of a search query for performance analysis.
     * @param index - Name of the index to profile query against
     * @param query - The search query to profile
     * @param options - Optional parameters:
     *   - LIMITED: Collect limited timing information only
     *   - All options supported by FT.SEARCH command
     */
    PROFILESEARCH: PROFILE_SEARCH_1.default,
    /**
     * Profiles the execution of a search query for performance analysis.
     * @param index - Name of the index to profile query against
     * @param query - The search query to profile
     * @param options - Optional parameters:
     *   - LIMITED: Collect limited timing information only
     *   - All options supported by FT.SEARCH command
     */
    profileSearch: PROFILE_SEARCH_1.default,
    /**
     * Profiles the execution of an aggregation query for performance analysis.
     * @param index - Name of the index to profile query against
     * @param query - The aggregation query to profile
     * @param options - Optional parameters:
     *   - LIMITED: Collect limited timing information only
     *   - All options supported by FT.AGGREGATE command
     */
    PROFILEAGGREGATE: PROFILE_AGGREGATE_1.default,
    /**
     * Profiles the execution of an aggregation query for performance analysis.
     * @param index - Name of the index to profile query against
     * @param query - The aggregation query to profile
     * @param options - Optional parameters:
     *   - LIMITED: Collect limited timing information only
     *   - All options supported by FT.AGGREGATE command
     */
    profileAggregate: PROFILE_AGGREGATE_1.default,
    /**
     * Performs a search query but returns only document ids without their contents.
     * @param args - Same parameters as FT.SEARCH:
     *   - parser: The command parser
     *   - index: Name of the index to search
     *   - query: The text query to search
     *   - options: Optional search parameters
     */
    SEARCH_NOCONTENT: SEARCH_NOCONTENT_1.default,
    /**
     * Performs a search query but returns only document ids without their contents.
     * @param args - Same parameters as FT.SEARCH:
     *   - parser: The command parser
     *   - index: Name of the index to search
     *   - query: The text query to search
     *   - options: Optional search parameters
     */
    searchNoContent: SEARCH_NOCONTENT_1.default,
    /**
     * Searches a RediSearch index with the given query.
     * @param index - The index name to search
     * @param query - The text query to search. For syntax, see https://redis.io/docs/stack/search/reference/query_syntax
     * @param options - Optional search parameters including:
     *   - VERBATIM: do not try to use stemming for query expansion
     *   - NOSTOPWORDS: do not filter stopwords from the query
     *   - INKEYS/INFIELDS: restrict the search to specific keys/fields
     *   - RETURN: limit which fields are returned
     *   - SUMMARIZE/HIGHLIGHT: create search result highlights
     *   - LIMIT: pagination control
     *   - SORTBY: sort results by a specific field
     *   - PARAMS: bind parameters to the query
     */
    SEARCH: SEARCH_1.default,
    /**
     * Searches a RediSearch index with the given query.
     * @param index - The index name to search
     * @param query - The text query to search. For syntax, see https://redis.io/docs/stack/search/reference/query_syntax
     * @param options - Optional search parameters including:
     *   - VERBATIM: do not try to use stemming for query expansion
     *   - NOSTOPWORDS: do not filter stopwords from the query
     *   - INKEYS/INFIELDS: restrict the search to specific keys/fields
     *   - RETURN: limit which fields are returned
     *   - SUMMARIZE/HIGHLIGHT: create search result highlights
     *   - LIMIT: pagination control
     *   - SORTBY: sort results by a specific field
     *   - PARAMS: bind parameters to the query
     */
    search: SEARCH_1.default,
    /**
     * Performs spelling correction on a search query.
     * @param index - Name of the index to use for spelling corrections
     * @param query - The search query to check for spelling
     * @param options - Optional parameters:
     *   - DISTANCE: Maximum Levenshtein distance for spelling suggestions
     *   - TERMS: Custom dictionary terms to include/exclude
     *   - DIALECT: Version of query dialect to use (defaults to 1)
     */
    SPELLCHECK: SPELLCHECK_1.default,
    /**
     * Performs spelling correction on a search query.
     * @param index - Name of the index to use for spelling corrections
     * @param query - The search query to check for spelling
     * @param options - Optional parameters:
     *   - DISTANCE: Maximum Levenshtein distance for spelling suggestions
     *   - TERMS: Custom dictionary terms to include/exclude
     *   - DIALECT: Version of query dialect to use (defaults to 1)
     */
    spellCheck: SPELLCHECK_1.default,
    /**
     * Adds a suggestion string to an auto-complete suggestion dictionary.
     * @param key - The suggestion dictionary key
     * @param string - The suggestion string to add
     * @param score - The suggestion score used for sorting
     * @param options - Optional parameters:
     *   - INCR: If true, increment the existing entry's score
     *   - PAYLOAD: Optional payload to associate with the suggestion
     */
    SUGADD: SUGADD_1.default,
    /**
     * Adds a suggestion string to an auto-complete suggestion dictionary.
     * @param key - The suggestion dictionary key
     * @param string - The suggestion string to add
     * @param score - The suggestion score used for sorting
     * @param options - Optional parameters:
     *   - INCR: If true, increment the existing entry's score
     *   - PAYLOAD: Optional payload to associate with the suggestion
     */
    sugAdd: SUGADD_1.default,
    /**
     * Deletes a string from a suggestion dictionary.
     * @param key - The suggestion dictionary key
     * @param string - The suggestion string to delete
     */
    SUGDEL: SUGDEL_1.default,
    /**
     * Deletes a string from a suggestion dictionary.
     * @param key - The suggestion dictionary key
     * @param string - The suggestion string to delete
     */
    sugDel: SUGDEL_1.default,
    /**
     * Gets completion suggestions with their payloads from a suggestion dictionary.
     * @param args - Same parameters as FT.SUGGET:
     *   - parser: The command parser
     *   - key: The suggestion dictionary key
     *   - prefix: The prefix to get completion suggestions for
     *   - options: Optional parameters for fuzzy matching and max results
     */
    SUGGET_WITHPAYLOADS: SUGGET_WITHPAYLOADS_1.default,
    /**
     * Gets completion suggestions with their payloads from a suggestion dictionary.
     * @param args - Same parameters as FT.SUGGET:
     *   - parser: The command parser
     *   - key: The suggestion dictionary key
     *   - prefix: The prefix to get completion suggestions for
     *   - options: Optional parameters for fuzzy matching and max results
     */
    sugGetWithPayloads: SUGGET_WITHPAYLOADS_1.default,
    /**
     * Gets completion suggestions with their scores and payloads from a suggestion dictionary.
     * @param args - Same parameters as FT.SUGGET:
     *   - parser: The command parser
     *   - key: The suggestion dictionary key
     *   - prefix: The prefix to get completion suggestions for
     *   - options: Optional parameters for fuzzy matching and max results
     */
    SUGGET_WITHSCORES_WITHPAYLOADS: SUGGET_WITHSCORES_WITHPAYLOADS_1.default,
    /**
     * Gets completion suggestions with their scores and payloads from a suggestion dictionary.
     * @param args - Same parameters as FT.SUGGET:
     *   - parser: The command parser
     *   - key: The suggestion dictionary key
     *   - prefix: The prefix to get completion suggestions for
     *   - options: Optional parameters for fuzzy matching and max results
     */
    sugGetWithScoresWithPayloads: SUGGET_WITHSCORES_WITHPAYLOADS_1.default,
    /**
     * Gets completion suggestions with their scores from a suggestion dictionary.
     * @param args - Same parameters as FT.SUGGET:
     *   - parser: The command parser
     *   - key: The suggestion dictionary key
     *   - prefix: The prefix to get completion suggestions for
     *   - options: Optional parameters for fuzzy matching and max results
     */
    SUGGET_WITHSCORES: SUGGET_WITHSCORES_1.default,
    /**
     * Gets completion suggestions with their scores from a suggestion dictionary.
     * @param args - Same parameters as FT.SUGGET:
     *   - parser: The command parser
     *   - key: The suggestion dictionary key
     *   - prefix: The prefix to get completion suggestions for
     *   - options: Optional parameters for fuzzy matching and max results
     */
    sugGetWithScores: SUGGET_WITHSCORES_1.default,
    /**
     * Gets completion suggestions for a prefix from a suggestion dictionary.
     * @param key - The suggestion dictionary key
     * @param prefix - The prefix to get completion suggestions for
     * @param options - Optional parameters:
     *   - FUZZY: Enable fuzzy prefix matching
     *   - MAX: Maximum number of results to return
     */
    SUGGET: SUGGET_1.default,
    /**
     * Gets completion suggestions for a prefix from a suggestion dictionary.
     * @param key - The suggestion dictionary key
     * @param prefix - The prefix to get completion suggestions for
     * @param options - Optional parameters:
     *   - FUZZY: Enable fuzzy prefix matching
     *   - MAX: Maximum number of results to return
     */
    sugGet: SUGGET_1.default,
    /**
     * Gets the size of a suggestion dictionary.
     * @param key - The suggestion dictionary key
     */
    SUGLEN: SUGLEN_1.default,
    /**
     * Gets the size of a suggestion dictionary.
     * @param key - The suggestion dictionary key
     */
    sugLen: SUGLEN_1.default,
    /**
     * Dumps the contents of a synonym group.
     * @param index - Name of the index that contains the synonym group
     */
    SYNDUMP: SYNDUMP_1.default,
    /**
     * Dumps the contents of a synonym group.
     * @param index - Name of the index that contains the synonym group
     */
    synDump: SYNDUMP_1.default,
    /**
     * Updates a synonym group with new terms.
     * @param index - Name of the index that contains the synonym group
     * @param groupId - ID of the synonym group to update
     * @param terms - One or more synonym terms to add to the group
     * @param options - Optional parameters:
     *   - SKIPINITIALSCAN: Skip the initial scan for existing documents
     */
    SYNUPDATE: SYNUPDATE_1.default,
    /**
     * Updates a synonym group with new terms.
     * @param index - Name of the index that contains the synonym group
     * @param groupId - ID of the synonym group to update
     * @param terms - One or more synonym terms to add to the group
     * @param options - Optional parameters:
     *   - SKIPINITIALSCAN: Skip the initial scan for existing documents
     */
    synUpdate: SYNUPDATE_1.default,
    /**
     * Returns the distinct values in a TAG field.
     * @param index - Name of the index
     * @param fieldName - Name of the TAG field to get values from
     */
    TAGVALS: TAGVALS_1.default,
    /**
     * Returns the distinct values in a TAG field.
     * @param index - Name of the index
     * @param fieldName - Name of the TAG field to get values from
     */
    tagVals: TAGVALS_1.default
};
