// @redis/search/dist/lib/commands/PROFILE_SEARCH.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformProfileReply = exports.extractProfileResultsReply = void 0;
const SEARCH_1 = __importStar(require("./SEARCH"));
const reply_transformers_1 = require("./reply-transformers");
function extractProfileResultsReply(reply) {
    const replyObject = (0, reply_transformers_1.mapLikeToObject)(reply);
    // Redis 8+ wraps results under `Results`.
    if (Object.hasOwn(replyObject, 'Results')) {
        return replyObject['Results'];
    }
    // Redis 7.4 RESP3 returns search/aggregate payload directly at top-level.
    if ((Object.hasOwn(replyObject, 'total_results') || Object.hasOwn(replyObject, 'total')) &&
        Object.hasOwn(replyObject, 'results')) {
        return reply;
    }
    if (Object.hasOwn(replyObject, 'results')) {
        return replyObject['results'];
    }
    return ((0, reply_transformers_1.getMapValue)(replyObject, ['results']) ?? reply);
}
exports.extractProfileResultsReply = extractProfileResultsReply;
function normalizeLegacyProfileReply(profile) {
    return (0, reply_transformers_1.mapLikeEntries)(profile).map(([key, value]) => {
        // Redis 7.4 often wraps iterator profiles as a single-element array containing an object.
        // Tests expect the inner object normalized directly as a flat key/value list.
        if (Array.isArray(value) && value.length === 1) {
            const first = value[0];
            if (Object.keys((0, reply_transformers_1.mapLikeToObject)(first)).length > 0) {
                return [key, (0, reply_transformers_1.normalizeProfileReply)(first)];
            }
        }
        return [key, (0, reply_transformers_1.normalizeProfileReply)(value)];
    });
}
function transformProfileReply(reply) {
    const replyObject = (0, reply_transformers_1.mapLikeToObject)(reply);
    const profile = (Object.hasOwn(replyObject, 'Profile') ?
        replyObject['Profile'] :
        Object.hasOwn(replyObject, 'profile') ?
            replyObject['profile'] :
            (0, reply_transformers_1.getMapValue)(replyObject, ['Profile', 'profile']));
    const profileObject = (0, reply_transformers_1.mapLikeToObject)(profile);
    // Redis 7.2 - 7.4 profile payload is a plain map keyed by timing labels.
    if (Object.hasOwn(profileObject, 'Total profile time')) {
        return normalizeLegacyProfileReply(profile);
    }
    return (0, reply_transformers_1.normalizeProfileReply)(profile);
}
exports.transformProfileReply = transformProfileReply;
function transformProfileSearchReplyResp3(reply, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- matches TransformReply contract
preserve, typeMapping) {
    return {
        results: SEARCH_1.default.transformReply[3](extractProfileResultsReply(reply), preserve, typeMapping),
        profile: transformProfileReply(reply)
    };
}
exports.default = {
    NOT_KEYED_COMMAND: true,
    IS_READ_ONLY: true,
    parseCommand(parser, index, query, options) {
        parser.push('FT.PROFILE', index, 'SEARCH');
        if (options?.LIMITED) {
            parser.push('LIMITED');
        }
        parser.push('QUERY', query);
        (0, SEARCH_1.parseSearchOptions)(parser, options);
    },
    transformReply: {
        2: (reply, 
        // eslint-disable-next-line @typescript-eslint/no-explicit-any -- matches TransformReply contract
        preserve, typeMapping) => {
            return {
                results: SEARCH_1.default.transformReply[2](reply[0], preserve, typeMapping),
                profile: reply[1]
            };
        },
        3: transformProfileSearchReplyResp3
    },
};
