// @redis/search/dist/lib/commands/reply-transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAggregateResultRow = exports.parseSearchResultRow = exports.normalizeProfileReply = exports.parseDocumentValue = exports.toCompatObject = exports.getMapValue = exports.mapLikeToFlatArray = exports.mapLikeToObject = exports.mapLikeValues = exports.mapLikeEntries = exports.isPlainObject = void 0;
const reply_utils_1 = require("@redis/client/dist/lib/commands/reply-utils");
Object.defineProperty(exports, "isPlainObject", { enumerable: true, get: function () { return reply_utils_1.isPlainObject; } });
Object.defineProperty(exports, "mapLikeEntries", { enumerable: true, get: function () { return reply_utils_1.mapLikeEntries; } });
Object.defineProperty(exports, "mapLikeValues", { enumerable: true, get: function () { return reply_utils_1.mapLikeValues; } });
Object.defineProperty(exports, "mapLikeToObject", { enumerable: true, get: function () { return reply_utils_1.mapLikeToObject; } });
Object.defineProperty(exports, "mapLikeToFlatArray", { enumerable: true, get: function () { return reply_utils_1.mapLikeToFlatArray; } });
Object.defineProperty(exports, "getMapValue", { enumerable: true, get: function () { return reply_utils_1.getMapValue; } });
function toCompatObject(value) {
    const descriptors = {};
    for (const [key, entryValue] of Object.entries(value)) {
        descriptors[key] = {
            value: entryValue,
            configurable: true,
            enumerable: true,
            writable: true
        };
    }
    return Object.defineProperties({}, descriptors);
}
exports.toCompatObject = toCompatObject;
function assignDocumentField(target, key, value) {
    if (key === '$') {
        const json = value?.toString?.() ?? value;
        if (typeof json === 'string') {
            try {
                Object.assign(target, JSON.parse(json));
                return;
            }
            catch {
                // Fallback to setting the raw value below.
            }
        }
    }
    target[key] = value;
}
function parseDocumentValue(value) {
    const document = {};
    for (const [key, entryValue] of (0, reply_utils_1.mapLikeEntries)(value)) {
        assignDocumentField(document, key, entryValue);
    }
    return document;
}
exports.parseDocumentValue = parseDocumentValue;
function normalizeProfileValue(value) {
    if (Array.isArray(value)) {
        return value.map(normalizeProfileValue);
    }
    if (value instanceof Map || (0, reply_utils_1.isPlainObject)(value)) {
        const normalized = [];
        for (const [key, entryValue] of (0, reply_utils_1.mapLikeEntries)(value)) {
            normalized.push(key, normalizeProfileValue(entryValue));
        }
        return normalized;
    }
    return value;
}
function normalizeProfileReply(profile) {
    return normalizeProfileValue(profile);
}
exports.normalizeProfileReply = normalizeProfileReply;
function parseSearchResultRow(rawRow) {
    const row = (0, reply_utils_1.mapLikeToObject)(rawRow);
    const value = {};
    Object.assign(value, parseDocumentValue((0, reply_utils_1.getMapValue)(row, ['values'])));
    Object.assign(value, parseDocumentValue((0, reply_utils_1.getMapValue)(row, ['extra_attributes', 'extraAttributes'])));
    return {
        id: (0, reply_utils_1.getMapValue)(row, ['id', 'doc_id']),
        value: toCompatObject(value)
    };
}
exports.parseSearchResultRow = parseSearchResultRow;
function parseAggregateResultRow(rawRow) {
    const row = (0, reply_utils_1.mapLikeToObject)(rawRow);
    const result = {};
    Object.assign(result, parseDocumentValue((0, reply_utils_1.getMapValue)(row, ['values'])));
    Object.assign(result, parseDocumentValue((0, reply_utils_1.getMapValue)(row, ['extra_attributes', 'extraAttributes'])));
    for (const [key, value] of Object.entries(row)) {
        if (key === 'id' ||
            key === 'values' ||
            key.toLowerCase() === 'extra_attributes' ||
            key === 'extraAttributes') {
            continue;
        }
        if (!Object.hasOwn(result, key)) {
            result[key] = value;
        }
    }
    return toCompatObject(result);
}
exports.parseAggregateResultRow = parseAggregateResultRow;
