// @redis/search/dist/lib/commands/SEARCH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseSearchOptions = exports.parseParamsArgument = void 0;
const generic_transformers_1 = require("@redis/client/dist/lib/commands/generic-transformers");
const default_1 = require("../dialect/default");
const reply_transformers_1 = require("./reply-transformers");
function parseParamsArgument(parser, params) {
    if (params) {
        parser.push('PARAMS');
        const args = [];
        for (const key in params) {
            if (!Object.hasOwn(params, key))
                continue;
            const value = params[key];
            args.push(key, typeof value === 'number' ? value.toString() : value);
        }
        parser.pushVariadicWithLength(args);
    }
}
exports.parseParamsArgument = parseParamsArgument;
function parseSearchOptions(parser, options) {
    if (options?.VERBATIM) {
        parser.push('VERBATIM');
    }
    if (options?.NOSTOPWORDS) {
        parser.push('NOSTOPWORDS');
    }
    (0, generic_transformers_1.parseOptionalVariadicArgument)(parser, 'INKEYS', options?.INKEYS);
    (0, generic_transformers_1.parseOptionalVariadicArgument)(parser, 'INFIELDS', options?.INFIELDS);
    (0, generic_transformers_1.parseOptionalVariadicArgument)(parser, 'RETURN', options?.RETURN);
    if (options?.SUMMARIZE) {
        parser.push('SUMMARIZE');
        if (typeof options.SUMMARIZE === 'object') {
            (0, generic_transformers_1.parseOptionalVariadicArgument)(parser, 'FIELDS', options.SUMMARIZE.FIELDS);
            if (options.SUMMARIZE.FRAGS !== undefined) {
                parser.push('FRAGS', options.SUMMARIZE.FRAGS.toString());
            }
            if (options.SUMMARIZE.LEN !== undefined) {
                parser.push('LEN', options.SUMMARIZE.LEN.toString());
            }
            if (options.SUMMARIZE.SEPARATOR !== undefined) {
                parser.push('SEPARATOR', options.SUMMARIZE.SEPARATOR);
            }
        }
    }
    if (options?.HIGHLIGHT) {
        parser.push('HIGHLIGHT');
        if (typeof options.HIGHLIGHT === 'object') {
            (0, generic_transformers_1.parseOptionalVariadicArgument)(parser, 'FIELDS', options.HIGHLIGHT.FIELDS);
            if (options.HIGHLIGHT.TAGS) {
                parser.push('TAGS', options.HIGHLIGHT.TAGS.open, options.HIGHLIGHT.TAGS.close);
            }
        }
    }
    if (options?.SLOP !== undefined) {
        parser.push('SLOP', options.SLOP.toString());
    }
    if (options?.TIMEOUT !== undefined) {
        parser.push('TIMEOUT', options.TIMEOUT.toString());
    }
    if (options?.INORDER) {
        parser.push('INORDER');
    }
    if (options?.LANGUAGE) {
        parser.push('LANGUAGE', options.LANGUAGE);
    }
    if (options?.EXPANDER) {
        parser.push('EXPANDER', options.EXPANDER);
    }
    if (options?.SCORER) {
        parser.push('SCORER', options.SCORER);
    }
    if (options?.SORTBY) {
        parser.push('SORTBY');
        if (typeof options.SORTBY === 'string' || options.SORTBY instanceof Buffer) {
            parser.push(options.SORTBY);
        }
        else {
            parser.push(options.SORTBY.BY);
            if (options.SORTBY.DIRECTION) {
                parser.push(options.SORTBY.DIRECTION);
            }
        }
    }
    if (options?.LIMIT) {
        parser.push('LIMIT', options.LIMIT.from.toString(), options.LIMIT.size.toString());
    }
    parseParamsArgument(parser, options?.PARAMS);
    if (options?.DIALECT) {
        parser.push('DIALECT', options.DIALECT.toString());
    }
    else {
        parser.push('DIALECT', default_1.DEFAULT_DIALECT);
    }
}
exports.parseSearchOptions = parseSearchOptions;
function transformSearchReplyResp2(reply, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- matches TransformReply contract
_preserve, _typeMapping) {
    // if reply[2] is array, then we have content/documents. Otherwise, only ids
    const withoutDocuments = reply.length > 2 && !Array.isArray(reply[2]);
    const documents = [];
    let i = 1;
    while (i < reply.length) {
        documents.push({
            id: reply[i++],
            value: (withoutDocuments ? {} : documentValue(reply[i++]))
        });
    }
    return {
        total: reply[0],
        documents
    };
}
function transformSearchReplyResp3(rawReply, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- matches TransformReply contract
preserve, typeMapping) {
    if (Array.isArray(rawReply)) {
        return transformSearchReplyResp2(rawReply, preserve, typeMapping);
    }
    const reply = (0, reply_transformers_1.mapLikeToObject)(rawReply);
    const total = Number((0, reply_transformers_1.getMapValue)(reply, ['total_results', 'total']) ?? 0);
    const results = (0, reply_transformers_1.mapLikeValues)((0, reply_transformers_1.getMapValue)(reply, ['results', 'documents']) ?? []);
    const documents = results.map(result => {
        const { id, value } = (0, reply_transformers_1.parseSearchResultRow)(result);
        return {
            id: String(id?.toString?.() ?? id ?? ''),
            value: value
        };
    });
    return {
        total,
        documents
    };
}
exports.default = {
    NOT_KEYED_COMMAND: true,
    IS_READ_ONLY: true,
    parseCommand(parser, index, query, options) {
        parser.push('FT.SEARCH', index, query);
        parseSearchOptions(parser, options);
    },
    transformReply: {
        2: transformSearchReplyResp2,
        3: transformSearchReplyResp3
    },
};
function documentValue(tuples) {
    return (0, reply_transformers_1.parseDocumentValue)(tuples);
}
