// @redis/search/dist/lib/commands/SPELLCHECK.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const default_1 = require("../dialect/default");
const reply_transformers_1 = require("./reply-transformers");
function transformSpellCheckReplyResp3(rawReply) {
    const transformed = [];
    const results = (0, reply_transformers_1.getMapValue)(rawReply, ['results', 'Results']) ?? rawReply;
    for (const [term, rawSuggestions] of (0, reply_transformers_1.mapLikeEntries)(results)) {
        const suggestions = [];
        for (const rawSuggestion of (0, reply_transformers_1.mapLikeValues)(rawSuggestions)) {
            if (Array.isArray(rawSuggestion) && rawSuggestion.length >= 2) {
                const first = rawSuggestion[0];
                const second = rawSuggestion[1];
                const numericFirst = Number(first);
                if (!Number.isNaN(numericFirst)) {
                    suggestions.push({
                        score: numericFirst,
                        suggestion: second.toString()
                    });
                }
                else {
                    suggestions.push({
                        score: Number(second),
                        suggestion: first.toString()
                    });
                }
                continue;
            }
            const entries = (0, reply_transformers_1.mapLikeEntries)(rawSuggestion);
            if (entries.length === 0)
                continue;
            const [suggestion, score] = entries[0];
            suggestions.push({
                score: Number(score),
                suggestion
            });
        }
        transformed.push({
            term,
            suggestions
        });
    }
    return transformed;
}
exports.default = {
    NOT_KEYED_COMMAND: true,
    IS_READ_ONLY: true,
    parseCommand(parser, index, query, options) {
        parser.push('FT.SPELLCHECK', index, query);
        if (options?.DISTANCE) {
            parser.push('DISTANCE', options.DISTANCE.toString());
        }
        if (options?.TERMS) {
            if (Array.isArray(options.TERMS)) {
                for (const term of options.TERMS) {
                    parseTerms(parser, term);
                }
            }
            else {
                parseTerms(parser, options.TERMS);
            }
        }
        if (options?.DIALECT) {
            parser.push('DIALECT', options.DIALECT.toString());
        }
        else {
            parser.push('DIALECT', default_1.DEFAULT_DIALECT);
        }
    },
    transformReply: {
        2: (rawReply) => {
            return rawReply.map(([, term, suggestions]) => ({
                term,
                suggestions: suggestions.map(([score, suggestion]) => ({
                    score: Number(score),
                    suggestion
                }))
            }));
        },
        3: transformSpellCheckReplyResp3,
    },
};
function parseTerms(parser, { mode, dictionary }) {
    parser.push('TERMS', mode, dictionary);
}
