// @redis/time-series/dist/lib/commands/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ADD_1 = __importDefault(require("./ADD"));
const ALTER_1 = __importDefault(require("./ALTER"));
const CREATE_1 = __importDefault(require("./CREATE"));
const CREATERULE_1 = __importDefault(require("./CREATERULE"));
const DECRBY_1 = __importDefault(require("./DECRBY"));
const DEL_1 = __importDefault(require("./DEL"));
const DELETERULE_1 = __importDefault(require("./DELETERULE"));
const GET_1 = __importDefault(require("./GET"));
const INCRBY_1 = __importDefault(require("./INCRBY"));
const INFO_DEBUG_1 = __importDefault(require("./INFO_DEBUG"));
const INFO_1 = __importDefault(require("./INFO"));
const MADD_1 = __importDefault(require("./MADD"));
const MGET_SELECTED_LABELS_1 = __importDefault(require("./MGET_SELECTED_LABELS"));
const MGET_WITHLABELS_1 = __importDefault(require("./MGET_WITHLABELS"));
const MGET_1 = __importDefault(require("./MGET"));
const MRANGE_GROUPBY_1 = __importDefault(require("./MRANGE_GROUPBY"));
const MRANGE_MULTIAGGR_1 = __importDefault(require("./MRANGE_MULTIAGGR"));
const MRANGE_SELECTED_LABELS_MULTIAGGR_1 = __importDefault(require("./MRANGE_SELECTED_LABELS_MULTIAGGR"));
const MRANGE_SELECTED_LABELS_GROUPBY_1 = __importDefault(require("./MRANGE_SELECTED_LABELS_GROUPBY"));
const MRANGE_SELECTED_LABELS_1 = __importDefault(require("./MRANGE_SELECTED_LABELS"));
const MRANGE_WITHLABELS_MULTIAGGR_1 = __importDefault(require("./MRANGE_WITHLABELS_MULTIAGGR"));
const MRANGE_WITHLABELS_GROUPBY_1 = __importDefault(require("./MRANGE_WITHLABELS_GROUPBY"));
const MRANGE_WITHLABELS_1 = __importDefault(require("./MRANGE_WITHLABELS"));
const MRANGE_1 = __importDefault(require("./MRANGE"));
const MREVRANGE_GROUPBY_1 = __importDefault(require("./MREVRANGE_GROUPBY"));
const MREVRANGE_MULTIAGGR_1 = __importDefault(require("./MREVRANGE_MULTIAGGR"));
const MREVRANGE_SELECTED_LABELS_MULTIAGGR_1 = __importDefault(require("./MREVRANGE_SELECTED_LABELS_MULTIAGGR"));
const MREVRANGE_SELECTED_LABELS_GROUPBY_1 = __importDefault(require("./MREVRANGE_SELECTED_LABELS_GROUPBY"));
const MREVRANGE_SELECTED_LABELS_1 = __importDefault(require("./MREVRANGE_SELECTED_LABELS"));
const MREVRANGE_WITHLABELS_MULTIAGGR_1 = __importDefault(require("./MREVRANGE_WITHLABELS_MULTIAGGR"));
const MREVRANGE_WITHLABELS_GROUPBY_1 = __importDefault(require("./MREVRANGE_WITHLABELS_GROUPBY"));
const MREVRANGE_WITHLABELS_1 = __importDefault(require("./MREVRANGE_WITHLABELS"));
const MREVRANGE_1 = __importDefault(require("./MREVRANGE"));
const QUERYINDEX_1 = __importDefault(require("./QUERYINDEX"));
const RANGE_MULTIAGGR_1 = __importDefault(require("./RANGE_MULTIAGGR"));
const RANGE_1 = __importDefault(require("./RANGE"));
const REVRANGE_MULTIAGGR_1 = __importDefault(require("./REVRANGE_MULTIAGGR"));
const REVRANGE_1 = __importDefault(require("./REVRANGE"));
__exportStar(require("./helpers"), exports);
exports.default = {
    /**
     * Creates or appends a sample to a time series
     * @param key - The key name for the time series
     * @param timestamp - The timestamp of the sample
     * @param value - The value of the sample
     * @param options - Optional configuration parameters
     */
    ADD: ADD_1.default,
    /**
     * Creates or appends a sample to a time series
     * @param key - The key name for the time series
     * @param timestamp - The timestamp of the sample
     * @param value - The value of the sample
     * @param options - Optional configuration parameters
     */
    add: ADD_1.default,
    /**
     * Alters the configuration of an existing time series
     * @param key - The key name for the time series
     * @param options - Configuration parameters to alter
     */
    ALTER: ALTER_1.default,
    /**
     * Alters the configuration of an existing time series
     * @param key - The key name for the time series
     * @param options - Configuration parameters to alter
     */
    alter: ALTER_1.default,
    /**
     * Creates a new time series
     * @param key - The key name for the new time series
     * @param options - Optional configuration parameters
     */
    CREATE: CREATE_1.default,
    /**
     * Creates a new time series
     * @param key - The key name for the new time series
     * @param options - Optional configuration parameters
     */
    create: CREATE_1.default,
    /**
     * Creates a compaction rule from source time series to destination time series
     * @param sourceKey - The source time series key
     * @param destinationKey - The destination time series key
     * @param aggregationType - The aggregation type to use
     * @param bucketDuration - The duration of each bucket in milliseconds
     * @param alignTimestamp - Optional timestamp for alignment
     */
    CREATERULE: CREATERULE_1.default,
    /**
     * Creates a compaction rule from source time series to destination time series
     * @param sourceKey - The source time series key
     * @param destinationKey - The destination time series key
     * @param aggregationType - The aggregation type to use
     * @param bucketDuration - The duration of each bucket in milliseconds
     * @param alignTimestamp - Optional timestamp for alignment
     */
    createRule: CREATERULE_1.default,
    /**
     * Decreases the value of a time series by a given amount
     * @param args - Arguments passed to the parseIncrByArguments function
     */
    DECRBY: DECRBY_1.default,
    /**
     * Decreases the value of a time series by a given amount
     * @param args - Arguments passed to the parseIncrByArguments function
     */
    decrBy: DECRBY_1.default,
    /**
     * Deletes samples between two timestamps from a time series
     * @param key - The key name of the time series
     * @param fromTimestamp - Start timestamp to delete from
     * @param toTimestamp - End timestamp to delete until
     */
    DEL: DEL_1.default,
    /**
     * Deletes samples between two timestamps from a time series
     * @param key - The key name of the time series
     * @param fromTimestamp - Start timestamp to delete from
     * @param toTimestamp - End timestamp to delete until
     */
    del: DEL_1.default,
    /**
     * Deletes a compaction rule between source and destination time series
     * @param sourceKey - The source time series key
     * @param destinationKey - The destination time series key
     */
    DELETERULE: DELETERULE_1.default,
    /**
     * Deletes a compaction rule between source and destination time series
     * @param sourceKey - The source time series key
     * @param destinationKey - The destination time series key
     */
    deleteRule: DELETERULE_1.default,
    /**
     * Gets the last sample of a time series
     * @param key - The key name of the time series
     * @param options - Optional parameters for the command
     */
    GET: GET_1.default,
    /**
     * Gets the last sample of a time series
     * @param key - The key name of the time series
     * @param options - Optional parameters for the command
     */
    get: GET_1.default,
    /**
     * Increases the value of a time series by a given amount
     * @param args - Arguments passed to the {@link parseIncrByArguments} function
     */
    INCRBY: INCRBY_1.default,
    /**
     * Increases the value of a time series by a given amount
     * @param args - Arguments passed to the {@link parseIncrByArguments} function
     */
    incrBy: INCRBY_1.default,
    /**
     * Gets debug information about a time series
     * @param key - The key name of the time series
     */
    INFO_DEBUG: INFO_DEBUG_1.default,
    /**
     * Gets debug information about a time series
     * @param key - The key name of the time series
     */
    infoDebug: INFO_DEBUG_1.default,
    /**
     * Gets information about a time series
     * @param key - The key name of the time series
     */
    INFO: INFO_1.default,
    /**
     * Gets information about a time series
     * @param key - The key name of the time series
     */
    info: INFO_1.default,
    /**
     * Adds multiple samples to multiple time series
     * @param toAdd - Array of samples to add to different time series
     */
    MADD: MADD_1.default,
    /**
     * Adds multiple samples to multiple time series
     * @param toAdd - Array of samples to add to different time series
     */
    mAdd: MADD_1.default,
    /**
     * Gets the last samples matching a specific filter with selected labels
     * @param filter - Filter to match time series keys
     * @param selectedLabels - Labels to include in the output
     * @param options - Optional parameters for the command
     */
    MGET_SELECTED_LABELS: MGET_SELECTED_LABELS_1.default,
    /**
     * Gets the last samples matching a specific filter with selected labels
     * @param filter - Filter to match time series keys
     * @param selectedLabels - Labels to include in the output
     * @param options - Optional parameters for the command
     */
    mGetSelectedLabels: MGET_SELECTED_LABELS_1.default,
    /**
     * Gets the last samples matching a specific filter with labels
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MGET_WITHLABELS: MGET_WITHLABELS_1.default,
    /**
     * Gets the last samples matching a specific filter with labels
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mGetWithLabels: MGET_WITHLABELS_1.default,
    /**
     * Gets the last samples matching a specific filter from multiple time series
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MGET: MGET_1.default,
    /**
     * Gets the last samples matching a specific filter from multiple time series
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mGet: MGET_1.default,
    /**
     * Gets samples for time series matching a filter within a time range with grouping
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    MRANGE_GROUPBY: MRANGE_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter within a time range with grouping
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    mRangeGroupBy: MRANGE_GROUPBY_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a specific filter within a time range
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MRANGE_MULTIAGGR: MRANGE_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a specific filter within a time range
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRangeMultiAggr: MRANGE_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with selected labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MRANGE_SELECTED_LABELS_MULTIAGGR: MRANGE_SELECTED_LABELS_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with selected labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRangeSelectedLabelsMultiAggr: MRANGE_SELECTED_LABELS_MULTIAGGR_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels and grouping
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    MRANGE_SELECTED_LABELS_GROUPBY: MRANGE_SELECTED_LABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels and grouping
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    mRangeSelectedLabelsGroupBy: MRANGE_SELECTED_LABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MRANGE_SELECTED_LABELS: MRANGE_SELECTED_LABELS_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRangeSelectedLabels: MRANGE_SELECTED_LABELS_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MRANGE_WITHLABELS_MULTIAGGR: MRANGE_WITHLABELS_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRangeWithLabelsMultiAggr: MRANGE_WITHLABELS_MULTIAGGR_1.default,
    /**
     * Gets samples for time series matching a filter with labels and grouping
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    MRANGE_WITHLABELS_GROUPBY: MRANGE_WITHLABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with labels and grouping
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    mRangeWithLabelsGroupBy: MRANGE_WITHLABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MRANGE_WITHLABELS: MRANGE_WITHLABELS_1.default,
    /**
     * Gets samples for time series matching a filter with labels
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRangeWithLabels: MRANGE_WITHLABELS_1.default,
    /**
     * Gets samples for time series matching a specific filter within a time range
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MRANGE: MRANGE_1.default,
    /**
     * Gets samples for time series matching a specific filter within a time range
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRange: MRANGE_1.default,
    /**
     * Gets samples for time series matching a filter within a time range with grouping (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    MREVRANGE_GROUPBY: MREVRANGE_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter within a time range with grouping (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    mRevRangeGroupBy: MREVRANGE_GROUPBY_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a specific filter within a time range (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MREVRANGE_MULTIAGGR: MREVRANGE_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a specific filter within a time range (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRevRangeMultiAggr: MREVRANGE_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with selected labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MREVRANGE_SELECTED_LABELS_MULTIAGGR: MREVRANGE_SELECTED_LABELS_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with selected labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRevRangeSelectedLabelsMultiAggr: MREVRANGE_SELECTED_LABELS_MULTIAGGR_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels and grouping (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    MREVRANGE_SELECTED_LABELS_GROUPBY: MREVRANGE_SELECTED_LABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels and grouping (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    mRevRangeSelectedLabelsGroupBy: MREVRANGE_SELECTED_LABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MREVRANGE_SELECTED_LABELS: MREVRANGE_SELECTED_LABELS_1.default,
    /**
     * Gets samples for time series matching a filter with selected labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param selectedLabels - Labels to include in the output
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRevRangeSelectedLabels: MREVRANGE_SELECTED_LABELS_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MREVRANGE_WITHLABELS_MULTIAGGR: MREVRANGE_WITHLABELS_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples for time series matching a filter with labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRevRangeWithLabelsMultiAggr: MREVRANGE_WITHLABELS_MULTIAGGR_1.default,
    /**
     * Gets samples for time series matching a filter with labels and grouping (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    MREVRANGE_WITHLABELS_GROUPBY: MREVRANGE_WITHLABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with labels and grouping (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param groupBy - Group by parameters
     * @param options - Optional parameters for the command
     */
    mRevRangeWithLabelsGroupBy: MREVRANGE_WITHLABELS_GROUPBY_1.default,
    /**
     * Gets samples for time series matching a filter with labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MREVRANGE_WITHLABELS: MREVRANGE_WITHLABELS_1.default,
    /**
     * Gets samples for time series matching a filter with labels (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRevRangeWithLabels: MREVRANGE_WITHLABELS_1.default,
    /**
     * Gets samples for time series matching a specific filter within a time range (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    MREVRANGE: MREVRANGE_1.default,
    /**
     * Gets samples for time series matching a specific filter within a time range (in reverse order)
     * @param fromTimestamp - Start timestamp for range
     * @param toTimestamp - End timestamp for range
     * @param filter - Filter to match time series keys
     * @param options - Optional parameters for the command
     */
    mRevRange: MREVRANGE_1.default,
    /**
     * Queries the index for time series matching a specific filter
     * @param filter - Filter to match time series labels
     */
    QUERYINDEX: QUERYINDEX_1.default,
    /**
     * Queries the index for time series matching a specific filter
     * @param filter - Filter to match time series labels
     */
    queryIndex: QUERYINDEX_1.default,
    /**
     * Gets multi-aggregation samples from a time series within a time range
     * @param args - Arguments passed to the {@link transformRangeMultiArguments} function
     */
    RANGE_MULTIAGGR: RANGE_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples from a time series within a time range
     * @param args - Arguments passed to the {@link transformRangeMultiArguments} function
     */
    rangeMultiAggr: RANGE_MULTIAGGR_1.default,
    /**
     * Gets samples from a time series within a time range
     * @param args - Arguments passed to the {@link transformRangeArguments} function
     */
    RANGE: RANGE_1.default,
    /**
     * Gets samples from a time series within a time range
     * @param args - Arguments passed to the {@link transformRangeArguments} function
     */
    range: RANGE_1.default,
    /**
     * Gets multi-aggregation samples from a time series within a time range (in reverse order)
     * @param args - Arguments passed to the {@link transformRangeMultiArguments} function
     */
    REVRANGE_MULTIAGGR: REVRANGE_MULTIAGGR_1.default,
    /**
     * Gets multi-aggregation samples from a time series within a time range (in reverse order)
     * @param args - Arguments passed to the {@link transformRangeMultiArguments} function
     */
    revRangeMultiAggr: REVRANGE_MULTIAGGR_1.default,
    /**
     * Gets samples from a time series within a time range (in reverse order)
     * @param args - Arguments passed to the {@link transformRangeArguments} function
     */
    REVRANGE: REVRANGE_1.default,
    /**
     * Gets samples from a time series within a time range (in reverse order)
     * @param args - Arguments passed to the {@link transformRangeArguments} function
     */
    revRange: REVRANGE_1.default
};
