// @redis/time-series/dist/lib/commands/INFO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CREATERULE_1 = require("./CREATERULE");
const generic_transformers_1 = require("@redis/client/dist/lib/commands/generic-transformers");
const reply_utils_1 = require("@redis/client/dist/lib/commands/reply-utils");
function normalizeInfoLabels(labels) {
    if (Array.isArray(labels)) {
        if (labels.every(item => Array.isArray(item) && item.length >= 2)) {
            return labels.map(item => [item[0], item[1]]);
        }
        const normalized = labels
            .map(label => {
            const object = (0, reply_utils_1.mapLikeToObject)(label);
            return [
                (0, reply_utils_1.getMapValue)(object, ['name', 'label']),
                (0, reply_utils_1.getMapValue)(object, ['value'])
            ];
        })
            .filter(([name]) => name !== undefined);
        if (normalized.length > 0) {
            return normalized;
        }
    }
    return (0, reply_utils_1.mapLikeEntries)(labels).map(([name, value]) => [name, value]);
}
function normalizeInfoRules(rules) {
    const normalized = [];
    const aggregationTypes = new Set(Object.values(CREATERULE_1.TIME_SERIES_AGGREGATION_TYPE).map(type => type.toUpperCase()));
    const parseRuleTuple = (rule) => {
        const stringCandidates = rule.filter((value) => typeof value === 'string' || value instanceof Buffer);
        const numberCandidates = rule.filter((value) => typeof value === 'number');
        const aggregationCandidate = stringCandidates.find(value => {
            return aggregationTypes.has(value.toString().toUpperCase());
        });
        const keyCandidate = stringCandidates.find(value => value !== aggregationCandidate);
        return [
            (keyCandidate ?? rule[0]),
            (numberCandidates[0] ?? Number(rule[1])),
            (aggregationCandidate ?? rule[2])
        ];
    };
    if (!Array.isArray(rules)) {
        for (const [key, value] of (0, reply_utils_1.mapLikeEntries)(rules)) {
            if (Array.isArray(value)) {
                const timeBucket = value.find(item => typeof item === 'number') ?? Number(value[0]);
                const aggregationType = value.find(item => {
                    return (typeof item === 'string' || item instanceof Buffer) &&
                        aggregationTypes.has(item.toString().toUpperCase());
                }) ?? value[1];
                normalized.push([
                    key,
                    timeBucket,
                    aggregationType
                ]);
                continue;
            }
            const object = (0, reply_utils_1.mapLikeToObject)(value);
            const timeBucket = (0, reply_utils_1.getMapValue)(object, ['timeBucket', 'time_bucket']);
            const aggregationType = (0, reply_utils_1.getMapValue)(object, ['aggregationType', 'aggregation_type']);
            normalized.push([
                key,
                timeBucket,
                aggregationType
            ]);
        }
        return normalized;
    }
    if (Array.isArray(rules) && rules.every(rule => Array.isArray(rule) && rule.length >= 3)) {
        return rules.map(rule => parseRuleTuple(rule));
    }
    for (const rule of (0, reply_utils_1.mapLikeValues)(rules)) {
        if (Array.isArray(rule)) {
            normalized.push(parseRuleTuple(rule));
            continue;
        }
        const object = (0, reply_utils_1.mapLikeToObject)(rule);
        const key = (0, reply_utils_1.getMapValue)(object, ['key']);
        const timeBucket = (0, reply_utils_1.getMapValue)(object, ['timeBucket', 'time_bucket']);
        const aggregationType = (0, reply_utils_1.getMapValue)(object, ['aggregationType', 'aggregation_type']);
        normalized.push(parseRuleTuple([key, timeBucket, aggregationType]));
    }
    return normalized;
}
function normalizeInfoRawReply(reply) {
    if (Array.isArray(reply)) {
        return reply;
    }
    const normalized = [];
    for (const [key, value] of (0, reply_utils_1.mapLikeEntries)(reply)) {
        switch (key) {
            case 'labels':
                normalized.push(key, normalizeInfoLabels(value));
                break;
            case 'rules':
                normalized.push(key, normalizeInfoRules(value));
                break;
            default:
                normalized.push(key, value);
                break;
        }
    }
    return normalized;
}
function transformInfoReplyResp2(reply, _, typeMapping) {
    const ret = {};
    for (let i = 0; i < reply.length; i += 2) {
        const key = reply[i].toString();
        switch (key) {
            case 'totalSamples':
            case 'memoryUsage':
            case 'firstTimestamp':
            case 'lastTimestamp':
            case 'retentionTime':
            case 'chunkCount':
            case 'chunkSize':
            case 'chunkType':
            case 'duplicatePolicy':
            case 'sourceKey':
            case 'ignoreMaxTimeDiff':
                ret[key] = reply[i + 1];
                break;
            case 'labels':
                ret[key] = reply[i + 1].map(([name, value]) => ({
                    name,
                    value
                }));
                break;
            case 'rules':
                ret[key] = reply[i + 1].map(([key, timeBucket, aggregationType]) => ({
                    key,
                    timeBucket,
                    aggregationType
                }));
                break;
            case 'ignoreMaxValDiff':
                ret[key] = generic_transformers_1.transformDoubleReply[2](reply[i + 1], undefined, typeMapping);
                break;
        }
    }
    return ret;
}
function transformInfoReplyResp3(reply, preserve, typeMapping) {
    return transformInfoReplyResp2(normalizeInfoRawReply(reply), preserve, typeMapping);
}
exports.default = {
    IS_READ_ONLY: true,
    parseCommand(parser, key) {
        parser.push('TS.INFO');
        parser.pushKey(key);
    },
    transformReply: {
        2: transformInfoReplyResp2,
        3: transformInfoReplyResp3
    },
};
