// @redis/time-series/dist/lib/commands/INFO_DEBUG.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const reply_utils_1 = require("@redis/client/dist/lib/commands/reply-utils");
const INFO_1 = __importDefault(require("./INFO"));
function normalizeChunks(chunks) {
    return (0, reply_utils_1.mapLikeValues)(chunks).map(chunk => {
        if (Array.isArray(chunk)) {
            if (chunk.length >= 10 && chunk[0] === 'startTimestamp') {
                return {
                    startTimestamp: chunk[1],
                    endTimestamp: chunk[3],
                    samples: chunk[5],
                    size: chunk[7],
                    bytesPerSample: chunk[9].toString()
                };
            }
            return {
                startTimestamp: chunk[0],
                endTimestamp: chunk[1],
                samples: chunk[2],
                size: chunk[3],
                bytesPerSample: chunk[4].toString()
            };
        }
        const object = (0, reply_utils_1.mapLikeToObject)(chunk);
        return {
            startTimestamp: object.startTimestamp ?? object.start_timestamp,
            endTimestamp: object.endTimestamp ?? object.end_timestamp,
            samples: object.samples,
            size: object.size,
            bytesPerSample: (object.bytesPerSample ?? object.bytes_per_sample).toString()
        };
    });
}
exports.default = {
    IS_READ_ONLY: INFO_1.default.IS_READ_ONLY,
    parseCommand(parser, key) {
        INFO_1.default.parseCommand(parser, key);
        parser.push('DEBUG');
    },
    transformReply: {
        2: (reply, _, typeMapping) => {
            const ret = INFO_1.default.transformReply[2](reply, _, typeMapping);
            for (let i = 0; i < reply.length; i += 2) {
                const key = reply[i].toString();
                switch (key) {
                    case 'keySelfName': {
                        ret[key] = reply[i + 1];
                        break;
                    }
                    case 'Chunks': {
                        ret['chunks'] = reply[i + 1].map(chunk => ({
                            startTimestamp: chunk[1],
                            endTimestamp: chunk[3],
                            samples: chunk[5],
                            size: chunk[7],
                            bytesPerSample: chunk[9]
                        }));
                        break;
                    }
                }
            }
            return ret;
        },
        3: (reply, preserve, typeMapping) => {
            const ret = INFO_1.default.transformReply[3](reply, preserve, typeMapping);
            const mappedReply = (0, reply_utils_1.mapLikeToObject)(reply);
            ret.keySelfName = (mappedReply.keySelfName ?? mappedReply.key_self_name);
            const chunks = mappedReply.Chunks ?? mappedReply.chunks;
            ret.chunks = normalizeChunks(chunks);
            return ret;
        }
    },
};
