// @redis/time-series/dist/lib/commands/MRANGE_WITHLABELS_MULTIAGGR.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTransformMRangeWithLabelsMultiArguments = void 0;
const helpers_1 = require("./helpers");
const RANGE_MULTIAGGR_1 = require("./RANGE_MULTIAGGR");
const MGET_1 = require("./MGET");
/**
 * Creates a function that parses arguments for multi-range commands with labels and multiple aggregators
 * @param command - The command name to use (TS.MRANGE or TS.MREVRANGE)
 */
function createTransformMRangeWithLabelsMultiArguments(command) {
    return (parser, fromTimestamp, toTimestamp, filter, options) => {
        parser.push(command);
        (0, RANGE_MULTIAGGR_1.parseRangeMultiArguments)(parser, fromTimestamp, toTimestamp, options);
        parser.push('WITHLABELS');
        (0, MGET_1.parseFilterArgument)(parser, filter);
    };
}
exports.createTransformMRangeWithLabelsMultiArguments = createTransformMRangeWithLabelsMultiArguments;
exports.default = {
    NOT_KEYED_COMMAND: true,
    IS_READ_ONLY: true,
    parseCommand: createTransformMRangeWithLabelsMultiArguments('TS.MRANGE'),
    transformReply: {
        2(reply, _, typeMapping) {
            return (0, helpers_1.resp2MapToValue)(reply, ([_key, labels, samples]) => {
                const unwrappedLabels = labels;
                // TODO: use Map type mapping for labels
                const labelsObject = {};
                for (const tuple of unwrappedLabels) {
                    const [key, value] = tuple;
                    const unwrappedKey = key;
                    labelsObject[unwrappedKey.toString()] = value;
                }
                return {
                    labels: labelsObject,
                    samples: helpers_1.transformMultiAggregationSamplesReply[2](samples)
                };
            }, typeMapping);
        },
        3(reply) {
            return (0, helpers_1.resp3MapToValue)(reply, ([labels, _metadata, samples]) => {
                return {
                    labels,
                    samples: helpers_1.transformMultiAggregationSamplesReply[3](samples)
                };
            });
        }
    },
};
