// @redis/client/dist/lib/commander.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scriptArgumentsPrefix = exports.functionArgumentsPrefix = exports.getTransformReply = exports.attachConfig = void 0;
const types_1 = require("./RESP/types");
function attachConfig({ BaseClass, commands, createCommand, createModuleCommand, createFunctionCommand, createScriptCommand, config }) {
    const RESP = config?.RESP ?? types_1.DEFAULT_RESP, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- dynamic prototype patching
    Class = class extends BaseClass {
    };
    for (const [name, command] of Object.entries(commands)) {
        Class.prototype[name] = createCommand(command, RESP);
    }
    if (config?.modules) {
        for (const [moduleName, module] of Object.entries(config.modules)) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- dynamic command namespace
            const fns = {};
            for (const [name, command] of Object.entries(module)) {
                fns[name] = createModuleCommand(command, RESP);
            }
            attachNamespace(Class.prototype, moduleName, fns);
        }
    }
    if (config?.functions) {
        for (const [library, commands] of Object.entries(config.functions)) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any -- dynamic command namespace
            const fns = {};
            for (const [name, command] of Object.entries(commands)) {
                fns[name] = createFunctionCommand(name, command, RESP);
            }
            attachNamespace(Class.prototype, library, fns);
        }
    }
    if (config?.scripts) {
        for (const [name, script] of Object.entries(config.scripts)) {
            Class.prototype[name] = createScriptCommand(script, RESP);
        }
    }
    return Class;
}
exports.attachConfig = attachConfig;
// Per-receiver namespace cache. Keyed by the receiver (original instance or any
// `withCommandOptions(...)` proxy) so each one gets a namespace bound to itself
// via `_self`. Caching the namespace as an own property on the receiver — which
// is what an earlier version did — leaks across the prototype chain: a proxy
// created via `Object.create(original)` would inherit the original's cached
// namespace and `_self` would point back to the original, silently bypassing
// the proxy's command-options overrides for every module/function command.
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- namespaces are dynamically shaped per module
const namespaceCache = new WeakMap();
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- dynamic prototype patching helper
function attachNamespace(prototype, name, fns) {
    Object.defineProperty(prototype, name, {
        get() {
            let perReceiver = namespaceCache.get(this);
            if (perReceiver === undefined) {
                perReceiver = new Map();
                namespaceCache.set(this, perReceiver);
            }
            let value = perReceiver.get(name);
            if (value === undefined) {
                value = Object.create(fns);
                value._self = this;
                perReceiver.set(name, value);
            }
            return value;
        }
    });
}
function getTransformReply(command, resp) {
    switch (typeof command.transformReply) {
        case 'function':
            return command.transformReply;
        case 'object':
            return command.transformReply[resp];
    }
}
exports.getTransformReply = getTransformReply;
function functionArgumentsPrefix(name, fn) {
    const prefix = [
        fn.IS_READ_ONLY ? 'FCALL_RO' : 'FCALL',
        name
    ];
    if (fn.NUMBER_OF_KEYS !== undefined) {
        prefix.push(fn.NUMBER_OF_KEYS.toString());
    }
    return prefix;
}
exports.functionArgumentsPrefix = functionArgumentsPrefix;
function scriptArgumentsPrefix(script) {
    const prefix = [
        script.IS_READ_ONLY ? 'EVALSHA_RO' : 'EVALSHA',
        script.SHA1
    ];
    if (script.NUMBER_OF_KEYS !== undefined) {
        prefix.push(script.NUMBER_OF_KEYS.toString());
    }
    return prefix;
}
exports.scriptArgumentsPrefix = scriptArgumentsPrefix;
