// @redis/json/dist/lib/commands/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformRedisJsonNullReply = exports.transformRedisJsonReply = exports.transformRedisJsonArgument = void 0;
const ARRAPPEND_1 = __importDefault(require("./ARRAPPEND"));
const ARRINDEX_1 = __importDefault(require("./ARRINDEX"));
const ARRINSERT_1 = __importDefault(require("./ARRINSERT"));
const ARRLEN_1 = __importDefault(require("./ARRLEN"));
const ARRPOP_1 = __importDefault(require("./ARRPOP"));
const ARRTRIM_1 = __importDefault(require("./ARRTRIM"));
const CLEAR_1 = __importDefault(require("./CLEAR"));
const DEBUG_MEMORY_1 = __importDefault(require("./DEBUG_MEMORY"));
const DEL_1 = __importDefault(require("./DEL"));
const FORGET_1 = __importDefault(require("./FORGET"));
const GET_1 = __importDefault(require("./GET"));
const MERGE_1 = __importDefault(require("./MERGE"));
const MGET_1 = __importDefault(require("./MGET"));
const MSET_1 = __importDefault(require("./MSET"));
const NUMINCRBY_1 = __importDefault(require("./NUMINCRBY"));
const NUMMULTBY_1 = __importDefault(require("./NUMMULTBY"));
const OBJKEYS_1 = __importDefault(require("./OBJKEYS"));
const OBJLEN_1 = __importDefault(require("./OBJLEN"));
// import RESP from './RESP';
const SET_1 = __importDefault(require("./SET"));
const STRAPPEND_1 = __importDefault(require("./STRAPPEND"));
const STRLEN_1 = __importDefault(require("./STRLEN"));
const TOGGLE_1 = __importDefault(require("./TOGGLE"));
const TYPE_1 = __importDefault(require("./TYPE"));
var generic_transformers_1 = require("@redis/client/dist/lib/commands/generic-transformers");
Object.defineProperty(exports, "transformRedisJsonArgument", { enumerable: true, get: function () { return generic_transformers_1.transformRedisJsonArgument; } });
Object.defineProperty(exports, "transformRedisJsonReply", { enumerable: true, get: function () { return generic_transformers_1.transformRedisJsonReply; } });
Object.defineProperty(exports, "transformRedisJsonNullReply", { enumerable: true, get: function () { return generic_transformers_1.transformRedisJsonNullReply; } });
exports.default = {
    /**
     * Appends one or more values to the end of an array in a JSON document.
     * Returns the new array length after append, or null if the path does not exist.
     *
     * @param key - The key to append to
     * @param path - Path to the array in the JSON document
     * @param json - The first value to append
     * @param jsons - Additional values to append
     */
    ARRAPPEND: ARRAPPEND_1.default,
    /**
     * Appends one or more values to the end of an array in a JSON document.
     * Returns the new array length after append, or null if the path does not exist.
     *
     * @param key - The key to append to
     * @param path - Path to the array in the JSON document
     * @param json - The first value to append
     * @param jsons - Additional values to append
     */
    arrAppend: ARRAPPEND_1.default,
    /**
     * Returns the index of the first occurrence of a value in a JSON array.
     * If the specified value is not found, it returns -1, or null if the path does not exist.
     *
     * @param key - The key containing the array
     * @param path - Path to the array in the JSON document
     * @param json - The value to search for
     * @param options - Optional range parameters for the search
     * @param options.range.start - Starting index for the search
     * @param options.range.stop - Optional ending index for the search
     */
    ARRINDEX: ARRINDEX_1.default,
    /**
     * Returns the index of the first occurrence of a value in a JSON array.
     * If the specified value is not found, it returns -1, or null if the path does not exist.
     *
     * @param key - The key containing the array
     * @param path - Path to the array in the JSON document
     * @param json - The value to search for
     * @param options - Optional range parameters for the search
     * @param options.range.start - Starting index for the search
     * @param options.range.stop - Optional ending index for the search
     */
    arrIndex: ARRINDEX_1.default,
    /**
     * Inserts one or more values into an array at the specified index.
     * Returns the new array length after insert, or null if the path does not exist.
     *
     * @param key - The key containing the array
     * @param path - Path to the array in the JSON document
     * @param index - The position where to insert the values
     * @param json - The first value to insert
     * @param jsons - Additional values to insert
     */
    ARRINSERT: ARRINSERT_1.default,
    /**
     * Inserts one or more values into an array at the specified index.
     * Returns the new array length after insert, or null if the path does not exist.
     *
     * @param key - The key containing the array
     * @param path - Path to the array in the JSON document
     * @param index - The position where to insert the values
     * @param json - The first value to insert
     * @param jsons - Additional values to insert
     */
    arrInsert: ARRINSERT_1.default,
    /**
     * Returns the length of an array in a JSON document.
     * Returns null if the path does not exist or the value is not an array.
     *
     * @param key - The key containing the array
     * @param options - Optional parameters
     * @param options.path - Path to the array in the JSON document
     */
    ARRLEN: ARRLEN_1.default,
    /**
     * Returns the length of an array in a JSON document.
     * Returns null if the path does not exist or the value is not an array.
     *
     * @param key - The key containing the array
     * @param options - Optional parameters
     * @param options.path - Path to the array in the JSON document
     */
    arrLen: ARRLEN_1.default,
    /**
     * Removes and returns an element from an array in a JSON document.
     * Returns null if the path does not exist or the value is not an array.
     *
     * @param key - The key containing the array
     * @param options - Optional parameters
     * @param options.path - Path to the array in the JSON document
     * @param options.index - Optional index to pop from. Default is -1 (last element)
     */
    ARRPOP: ARRPOP_1.default,
    /**
     * Removes and returns an element from an array in a JSON document.
     * Returns null if the path does not exist or the value is not an array.
     *
     * @param key - The key containing the array
     * @param options - Optional parameters
     * @param options.path - Path to the array in the JSON document
     * @param options.index - Optional index to pop from. Default is -1 (last element)
     */
    arrPop: ARRPOP_1.default,
    /**
     * Trims an array in a JSON document to include only elements within the specified range.
     * Returns the new array length after trimming, or null if the path does not exist.
     *
     * @param key - The key containing the array
     * @param path - Path to the array in the JSON document
     * @param start - Starting index (inclusive)
     * @param stop - Ending index (inclusive)
     */
    ARRTRIM: ARRTRIM_1.default,
    /**
     * Trims an array in a JSON document to include only elements within the specified range.
     * Returns the new array length after trimming, or null if the path does not exist.
     *
     * @param key - The key containing the array
     * @param path - Path to the array in the JSON document
     * @param start - Starting index (inclusive)
     * @param stop - Ending index (inclusive)
     */
    arrTrim: ARRTRIM_1.default,
    /**
     * Clears container values (arrays/objects) in a JSON document.
     * Returns the number of values cleared (0 or 1), or null if the path does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the container to clear
     */
    CLEAR: CLEAR_1.default,
    /**
     * Clears container values (arrays/objects) in a JSON document.
     * Returns the number of values cleared (0 or 1), or null if the path does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the container to clear
     */
    clear: CLEAR_1.default,
    /**
     * Reports memory usage details for a JSON document value.
     * Returns size in bytes of the value, or null if the key or path does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the value to examine
     */
    DEBUG_MEMORY: DEBUG_MEMORY_1.default,
    /**
     * Reports memory usage details for a JSON document value.
     * Returns size in bytes of the value, or null if the key or path does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the value to examine
     */
    debugMemory: DEBUG_MEMORY_1.default,
    /**
     * Deletes a value from a JSON document.
     * Returns the number of paths deleted (0 or 1), or null if the key does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the value to delete
     */
    DEL: DEL_1.default,
    /**
     * Deletes a value from a JSON document.
     * Returns the number of paths deleted (0 or 1), or null if the key does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the value to delete
     */
    del: DEL_1.default,
    /**
     * Alias for JSON.DEL - Deletes a value from a JSON document.
     * Returns the number of paths deleted (0 or 1), or null if the key does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the value to delete
     */
    FORGET: FORGET_1.default,
    /**
     * Alias for JSON.DEL - Deletes a value from a JSON document.
     * Returns the number of paths deleted (0 or 1), or null if the key does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the value to delete
     */
    forget: FORGET_1.default,
    /**
     * Gets values from a JSON document.
     * Returns the value at the specified path, or null if the key or path does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path(s) to the value(s) to retrieve
     */
    GET: GET_1.default,
    /**
     * Gets values from a JSON document.
     * Returns the value at the specified path, or null if the key or path does not exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path(s) to the value(s) to retrieve
     */
    get: GET_1.default,
    /**
     * Merges a given JSON value into a JSON document.
     * Returns OK on success, or null if the key does not exist.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to merge into
     * @param value - JSON value to merge
     */
    MERGE: MERGE_1.default,
    /**
     * Merges a given JSON value into a JSON document.
     * Returns OK on success, or null if the key does not exist.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to merge into
     * @param value - JSON value to merge
     */
    merge: MERGE_1.default,
    /**
     * Gets values at a specific path from multiple JSON documents.
     * Returns an array of values at the path from each key, null for missing keys/paths.
     *
     * @param keys - Array of keys containing JSON documents
     * @param path - Path to retrieve from each document
     */
    MGET: MGET_1.default,
    /**
     * Gets values at a specific path from multiple JSON documents.
     * Returns an array of values at the path from each key, null for missing keys/paths.
     *
     * @param keys - Array of keys containing JSON documents
     * @param path - Path to retrieve from each document
     */
    mGet: MGET_1.default,
    /**
     * Sets multiple JSON values in multiple documents.
     * Returns OK on success.
     *
     * @param items - Array of objects containing key, path, and value to set
     * @param items[].key - The key containing the JSON document
     * @param items[].path - Path in the document to set
     * @param items[].value - JSON value to set at the path
     */
    MSET: MSET_1.default,
    /**
     * Sets multiple JSON values in multiple documents.
     * Returns OK on success.
     *
     * @param items - Array of objects containing key, path, and value to set
     * @param items[].key - The key containing the JSON document
     * @param items[].path - Path in the document to set
     * @param items[].value - JSON value to set at the path
     */
    mSet: MSET_1.default,
    /**
     * Increments a numeric value stored in a JSON document by a given number.
     * Returns the value after increment, or null if the key/path doesn't exist or value is not numeric.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to the numeric value
     * @param by - Amount to increment by
     */
    NUMINCRBY: NUMINCRBY_1.default,
    /**
     * Increments a numeric value stored in a JSON document by a given number.
     * Returns the value after increment, or null if the key/path doesn't exist or value is not numeric.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to the numeric value
     * @param by - Amount to increment by
     */
    numIncrBy: NUMINCRBY_1.default,
    /**
     * Multiplies a numeric value stored in a JSON document by a given number.
     * Returns the value after multiplication, or null if the key/path doesn't exist or value is not numeric.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to the numeric value
     * @param by - Amount to multiply by
     */
    NUMMULTBY: NUMMULTBY_1.default,
    /**
     * Multiplies a numeric value stored in a JSON document by a given number.
     * Returns the value after multiplication, or null if the key/path doesn't exist or value is not numeric.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to the numeric value
     * @param by - Amount to multiply by
     */
    numMultBy: NUMMULTBY_1.default,
    /**
     * Returns the keys in the object stored in a JSON document.
     * Returns array of keys, array of arrays for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the object to examine
     */
    OBJKEYS: OBJKEYS_1.default,
    /**
     * Returns the keys in the object stored in a JSON document.
     * Returns array of keys, array of arrays for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the object to examine
     */
    objKeys: OBJKEYS_1.default,
    /**
     * Returns the number of keys in the object stored in a JSON document.
     * Returns length of object, array of lengths for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the object to examine
     */
    OBJLEN: OBJLEN_1.default,
    /**
     * Returns the number of keys in the object stored in a JSON document.
     * Returns length of object, array of lengths for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the object to examine
     */
    objLen: OBJLEN_1.default,
    // RESP,
    // resp: RESP,
    /**
     * Sets a JSON value at a specific path in a JSON document.
     * Returns OK on success, or null if condition (NX/XX) is not met.
     *
     * @param key - The key containing the JSON document
     * @param path - Path in the document to set
     * @param json - JSON value to set at the path
     * @param options - Optional parameters
     * @param options.condition - Set condition: NX (only if doesn't exist) or XX (only if exists)
     * @deprecated options.NX - Use options.condition instead
     * @deprecated options.XX - Use options.condition instead
     */
    SET: SET_1.default,
    /**
     * Sets a JSON value at a specific path in a JSON document.
     * Returns OK on success, or null if condition (NX/XX) is not met.
     *
     * @param key - The key containing the JSON document
     * @param path - Path in the document to set
     * @param json - JSON value to set at the path
     * @param options - Optional parameters
     * @param options.condition - Set condition: NX (only if doesn't exist) or XX (only if exists)
     * @deprecated options.NX - Use options.condition instead
     * @deprecated options.XX - Use options.condition instead
     */
    set: SET_1.default,
    /**
     * Appends a string to a string value stored in a JSON document.
     * Returns new string length after append, or null if the path doesn't exist or value is not a string.
     *
     * @param key - The key containing the JSON document
     * @param append - String to append
     * @param options - Optional parameters
     * @param options.path - Path to the string value
     */
    STRAPPEND: STRAPPEND_1.default,
    /**
     * Appends a string to a string value stored in a JSON document.
     * Returns new string length after append, or null if the path doesn't exist or value is not a string.
     *
     * @param key - The key containing the JSON document
     * @param append - String to append
     * @param options - Optional parameters
     * @param options.path - Path to the string value
     */
    strAppend: STRAPPEND_1.default,
    /**
     * Returns the length of a string value stored in a JSON document.
     * Returns string length, array of lengths for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the string value
     */
    STRLEN: STRLEN_1.default,
    /**
     * Returns the length of a string value stored in a JSON document.
     * Returns string length, array of lengths for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to the string value
     */
    strLen: STRLEN_1.default,
    /**
     * Toggles a boolean value stored in a JSON document.
     * Returns 1 if value was toggled to true, 0 if toggled to false, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to the boolean value
     */
    TOGGLE: TOGGLE_1.default,
    /**
     * Toggles a boolean value stored in a JSON document.
     * Returns 1 if value was toggled to true, 0 if toggled to false, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param path - Path to the boolean value
     */
    toggle: TOGGLE_1.default,
    /**
     * Returns the type of JSON value at a specific path in a JSON document.
     * Returns the type as a string, array of types for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to examine
     */
    TYPE: TYPE_1.default,
    /**
     * Returns the type of JSON value at a specific path in a JSON document.
     * Returns the type as a string, array of types for multiple paths, or null if path doesn't exist.
     *
     * @param key - The key containing the JSON document
     * @param options - Optional parameters
     * @param options.path - Path to examine
     */
    type: TYPE_1.default
};
