// @redis/client/dist/lib/commands/ARGREP.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArGrepArguments = exports.AR_PREDICATE_COMBINATORS = exports.AR_PREDICATE_TYPES = void 0;
exports.AR_PREDICATE_TYPES = {
    EXACT: 'EXACT',
    MATCH: 'MATCH',
    GLOB: 'GLOB',
    RE: 'RE'
};
exports.AR_PREDICATE_COMBINATORS = {
    AND: 'AND',
    OR: 'OR'
};
function parseArGrepArguments(parser, key, start, end, predicates, options) {
    parser.pushKey(key);
    parser.push(typeof start === 'number' ? start.toString() : start, typeof end === 'number' ? end.toString() : end);
    for (const [type, value] of predicates) {
        parser.push(type, value);
    }
    if (options?.COMBINATOR !== undefined) {
        parser.push(options.COMBINATOR);
    }
    if (options?.LIMIT !== undefined) {
        parser.push('LIMIT', options.LIMIT.toString());
    }
    if (options?.NOCASE) {
        parser.push('NOCASE');
    }
}
exports.parseArGrepArguments = parseArGrepArguments;
exports.default = {
    IS_READ_ONLY: true,
    parseCommand(parser, ...args) {
        parser.push('ARGREP');
        parseArGrepArguments(parser, ...args);
    },
    transformReply: undefined
};
