// @redis/client/dist/lib/commands/GEOSEARCH_WITH.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GEO_REPLY_WITH = void 0;
const decoder_1 = require("../RESP/decoder");
const GEOSEARCH_1 = __importDefault(require("./GEOSEARCH"));
const generic_transformers_1 = require("./generic-transformers");
exports.GEO_REPLY_WITH = {
    DISTANCE: 'WITHDIST',
    HASH: 'WITHHASH',
    COORDINATES: 'WITHCOORD'
};
exports.default = {
    IS_READ_ONLY: GEOSEARCH_1.default.IS_READ_ONLY,
    parseCommand(parser, key, from, by, replyWith, options) {
        GEOSEARCH_1.default.parseCommand(parser, key, from, by, options);
        parser.push(...replyWith);
        parser.preserve = replyWith;
    },
    transformReply(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- heterogeneous tuple variance marker
    reply, replyWith, typeMapping) {
        const replyWithSet = new Set(replyWith);
        let index = 0;
        const distanceIndex = replyWithSet.has(exports.GEO_REPLY_WITH.DISTANCE) && ++index, hashIndex = replyWithSet.has(exports.GEO_REPLY_WITH.HASH) && ++index, coordinatesIndex = replyWithSet.has(exports.GEO_REPLY_WITH.COORDINATES) && ++index;
        const doubleMapping = typeMapping ? typeMapping[decoder_1.RESP_TYPES.DOUBLE] : undefined;
        const parseDouble = (value) => {
            if (typeof value !== 'number') {
                return generic_transformers_1.transformDoubleReply[2](value, undefined, typeMapping);
            }
            if (doubleMapping === String) {
                return value.toString();
            }
            return value;
        };
        return reply.map(raw => {
            const unwrapped = raw;
            const item = {
                member: unwrapped[0]
            };
            if (distanceIndex) {
                item.distance = parseDouble(unwrapped[distanceIndex]);
            }
            if (hashIndex) {
                item.hash = unwrapped[hashIndex];
            }
            if (coordinatesIndex) {
                const [longitude, latitude] = unwrapped[coordinatesIndex];
                item.coordinates = {
                    longitude: parseDouble(longitude),
                    latitude: parseDouble(latitude)
                };
            }
            return item;
        });
    }
};
