// @redis/client/dist/lib/commands/MODULE_LIST.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function moduleEntries(moduleReply) {
    if (Array.isArray(moduleReply)) {
        const entries = [];
        for (let i = 0; i + 1 < moduleReply.length; i += 2) {
            const key = moduleReply[i];
            if (key === null || key === undefined)
                continue;
            entries.push([key.toString(), moduleReply[i + 1]]);
        }
        return entries;
    }
    if (moduleReply instanceof Map) {
        return Array.from(moduleReply.entries(), ([key, value]) => {
            const k = key === null || key === undefined
                ? ''
                : key.toString();
            return [k, value];
        });
    }
    if (moduleReply !== null && typeof moduleReply === 'object') {
        return Object.entries(moduleReply);
    }
    return [];
}
function transformModuleReply(moduleReply) {
    const result = {};
    for (const [key, value] of moduleEntries(moduleReply)) {
        result[key] = value;
    }
    return result;
}
function transformModuleListReply(reply) {
    return reply.map(moduleReply => transformModuleReply(moduleReply));
}
exports.default = {
    NOT_KEYED_COMMAND: true,
    IS_READ_ONLY: true,
    parseCommand(parser) {
        parser.push('MODULE', 'LIST');
    },
    transformReply: {
        2: transformModuleListReply,
        3: transformModuleListReply
    }
};
